plugins/module_utils/foreman_helper.py

```diff
@@ -306,9 +307,23 @@ def api_connection(self):
     def changed(self):
         return self._changed
 
+    def flatten_entity_params(self):
+        """ Flattens taxonimy and other parameters that are of entity or entity_list type.
+             Will work only after connect()
+        """
+        self.auto_lookup_entities()
+        self.foreman_params = _flatten_entity(self.foreman_params, self.foreman_spec)
+
```

Comment:
The taxonomy (and potentially other entity params) need lookup and flattening for non-entity modules.
This is the cheapest way I could think of.

So now I need to call this from the module after connect()
Is there are smart way of doing this? And should we make this done automatically?

Reply:
cc @evgeni

Reply:
rename to _lookup_entity_params and call it in connect() for ForemanTaxonomicAnsibleModule?

Reply:
So the problem here is, that we're not in an "entity" module, so `_flatten_entity` is never called (it's called from the various entity create/update funcs we have, but those are not called here for obvious reasons). And additionally, given this is not an entity module, the lookup also does not happen.

I am not sure I like that both these tasks are done in one function with a name only pointing at one part, but I see that those calls are required.

@mdellweg any elegant ideas how to solve that?

Reply:
Maybe i am not seeing the whole picture, but it seems to be quite special to the templates_import module. You could just call those two lines there.

Reply:
it's special for every non-entity, non-katello module that does reference entities, we just don't have many of those ;)

but yeah, I think the cleanest solution for now would be not having this function and use

```python
    with module.api_connection():

        module.auto_lookup_entities()

        # Build a list of all existing templates of all supported types to check if we are adding any new
        template_report = []

        template_types = ['provisioning_templates', 'report_templates', 'ptables']
        if 'job_templates' in module.foremanapi.resources:
            template_types.append('job_templates')

        for template_type in template_types:
            template_report += [(resource['name'], resource['id']) for resource in module.list_resource(template_type)]

        result = module.resource_action('templates', 'import', record_change=False, params=_flatten_entity(module.foreman_params, module.foreman_spec))
```

(well, and obviously importing `_flatten_entity` at the top

Reply:
or build the payload with scope
`payload = module.scope_for('location')`

Reply:
This will also automatically look up location.

Reply:
Gone evgeni's way for now. scope_for seems longer - it'll need looping thu orgs and locations, right?